### ISSUE TITLE
Make Database.Command logger available on MigrationCommandExecutor

### ISSUE THREAD

**Issue by archanasoni** [dotnet/efcore#20495]
Hi,
We are working to support .NET Core 3.1 with IBM Entity Framework Provider.
Due to some specific requirement for DB2, we have implemented IMigrationCommandExecutor
and Logger object we are obtaining via constructor injection:
public MyCommandExecutor(
            [NotNull] IDiagnosticsLogger<Microsoft.EntityFrameworkCore.DbLoggerCategory.Database.Command> logger          
            )
        {
...
...
}

Here we do get logger instance, but registered interceptors are not initialized, due to this when we call logger method CommandReaderExecuting() we hit error:
System.NullReferenceException: Object reference not set to an instance of an object.
   at Microsoft.EntityFrameworkCore.Diagnostics.Internal.Interceptors.get_CoreOptionsExtension() in C:\Users\asoni\Downloads\EntityFrameworkCore-3.1.0\EntityFrameworkCore-3.1.0\src\EFCore\Diagnostics\Internal\Interceptors.cs:line 80
   at Microsoft.EntityFrameworkCore.Diagnostics.Internal.Interceptors.get_RegisteredInterceptors() in C:\Users\asoni\Downloads\EntityFrameworkCore-3.1.0\EntityFrameworkCore-3.1.0\src\EFCore\Diagnostics\Internal\Interceptors.cs:line 50
   at Microsoft.EntityFrameworkCore.Diagnostics.Internal.Interceptors.Aggregate[TInterceptor]() in C:\Users\asoni\Downloads\EntityFrameworkCore-3.1.0\EntityFrameworkCore-3.1.0\src\EFCore\Diagnostics\Internal\Interceptors.cs:line 71
   at Microsoft.EntityFrameworkCore.Diagnostics.RelationalLoggerExtensions.CommandReaderExecuting(IDiagnosticsLogger`1 diagnostics, IRelationalConnection connection, DbCommand command, DbContext context, Guid commandId, Guid connectionId, DateTimeOffset startTime) in C:\Users\asoni\Downloads\EntityFrameworkCore-3.1.0\EntityFrameworkCore-3.1.0\src\EFCore.Relational\Diagnostics\RelationalLoggerExtensions.cs:line 280
....
...
..

Not only this, but we do have our own method to capture some more data using this logger, that also doesn't print anything. log level is also not coming properly even after registering logger factory in context class.

We registered service:
builder.TryAdd<IMigrationCommandExecutor, MyCommandExecutor>();

This worked fine till 2.1, But when we migrated to 3.1 we observed this issue...

Thanks,
Archana

**Comment by ajcvickers** [comment]
@archanasoni Check your service dependencies--this could be a singleton service depending on a scoped service.

**Comment by archanasoni** [comment]
@ajcvickers , We just have one dependency that is logger and this logger is also built-in one..
MyCommandExecutor(
[NotNull] IDiagnosticsLogger<Microsoft.EntityFrameworkCore.DbLoggerCategory.Database.Command> logger)

We wanted to have Database.Command logger object so used constructor injection..
this perfectly worked till 2.1, From 3.1 onward we are facing issue..

I tried to debug SQLServer scaffolding as well, there also in case of scaffolding I could see registered interceptors are not initialized.

We see logger instance of MigrationsSqlGeneratorDependencies, This is being initialized perfectly.

In simpler word our issue is- How can we get hold of Database.Command logger in MigrationCommandExecutor class, via constructor injection it's not proper..

**Comment by ajcvickers** [comment]
@archanasoni I don't think it's currently possible to get a logger in `MigrationCommandExecutor`. It would need to be passed in to `ExecuteNoQuery` or available on one of the arguments. `MigrationCommand` instances have the logger, but it's not exposed publicly.

/cc @bricelam

**Comment by archanasoni** [comment]
Thanks @ajcvickers , But we have our own extra method as well to capture some extra information..
Without getting logger, We won't be able to capture important information.

This worked in 2.1, some breaking changes went into 3.1.

When can we expect resolution for this ?

**Comment by archanasoni** [comment]
@ajcvickers, We've our implementation for MigrationCommandExecutor reasons are:
1. For some of the operation we needed to get Tablespace and other info from Database at runtime, Here Db2 for zOS doesn't support PLSQL, so first we had to get the info from database, then modify command text and execute, it's like multiple call for ExecuteReader and ExecuteNonQuery, to achieve the one task.. Now Migration command text is readonly we can not edit it, hence we need to deal with it in our own way and to call logger explicitly..

2. For some operation we needed to run builtin Procedure, and this procedures returns reader as output of executed command.
It's very important for us to capture result returned by built in procedure.. And to achieve that we had implemented our own method to capture it..

Since with Core 3.1, we don't get logger, we wont be able to capture extra logs which is important for us.. and due to this diagnosing customer issue will be difficult for us..
